### ISSUE TITLE
Excluding service in escalations using ! doesn't function properly

### ISSUE THREAD

**Issue by scottwilkerson** [NagiosEnterprises/nagioscore#677]
If I create an escalation like this
```
define serviceescalation {
    host_name                localhost
    service_description      *,!HTTP
    contacts                 nagiosadmin
    first_notification       0
    last_notification        1
    notification_interval    60
    escalation_period        24x7
    escalation_options       c,
}
```
It should include all services except HTTP. When verifying the configuration I have confirmed it shows 1 less escalation in the line "Checked XX service escalations." than before I added the exclusion to the service_description line.
`,!HTTP`

However it is still found in objects.cache
```
define serviceescalation {
        host_name       localhost
        service_description     HTTP
        first_notification      0
        last_notification       1
        notification_interval   60.000000
        escalation_period       24x7
        escalation_options      c
        contacts        nagiosadmin
        }
```

**Comment by ericloyd** [comment]
Out of curiosity, what does Configuration -> Service Dependencies show in the UI?

**Comment by scottwilkerson** [comment]
It's an escalation but ironically Configuration -> Service Escalations actually shows it twice instead of not at all

> localhost | HTTP | nagiosadmin | 0 | 1 | 1h 0m 0s | 24x7 | Critical
> localhost | HTTP | nagiosadmin | 0 | 1 | 1h 0m 0s | 24x7 | Critical

**Comment by Madlohe** [comment]
Going by the [documentation](https://assets.nagios.com/downloads/nagioscore/docs/nagioscore/4/en/objecttricks.html), it seems like the "exclusion" syntax is limited to hosts at the moment. I didn't see that syntax used on any other pages. The use case you give makes sense to me, though.

**Comment by ericloyd** [comment]
If only @scottwilkerson knew someone who could add that to the codebase...

**Comment by sawolf** [comment]
After stepping through the template parsing code with a debugger, I realized that I mistyped my configuration. Once corrected, I wasn't able to reproduce the objects.cache behavior or the duplicate entries in System->Configuration->Service Escalations. Here's the config I used:

```
define hostgroup {
    hostgroup_name _testgrp_1
    alias Test Host Group 1
    members _testhost_1,_testhost_2
}

define host {
    host_name _testhost_1
    address 127.0.0.1
    use xiwizard_generic_host
}

define host {
    host_name _testhost_2
    address 127.0.0.1
    use xiwizard_generic_host
}

define service {
        hostgroup_name  _testgrp_1
        service_description _testservice_ping
        use xiwizard_website_ping_service
}

define service {
        host_name *,!_testhost_2
        service_description _testservice_http 
        use xiwizard_website_http_service     
}

define serviceescalation {
    host_name               _testhost_1
    service_description     *,!_testservice_ping
    contacts                nagiosadmin
    first_notification      0
    last_notification       1
    notification_interval   60
    escalation_period       24x7
    escalation_options      c,
}
```
If you try to run this on a machine where XI isn't installed, you'll need to copy over a large-ish subset of the configs included by default with that product.

Adding and removing the `,!_testservice_ping` from the service escalation seems to affect objects.cache as it should. @scottwilkerson do you have a configuration which causes this issue consistently?

**Comment by sawolf** [comment]
Okay, I talked to Scott about this offline. I haven't managed to reproduce this issue in 4.4.5 or in XI 5.6.6 (which uses Core 4.4.3), either by manually adding the serviceescalation or by adding via the CCM and manually adding the exclusion before restarting.

@ericloyd, Scott mentions you originally reported this on the forums. If you see this issue again on a newer system, please let me know. I'm going to close this issue for now.

**Comment by ericloyd** [comment]
I'd have to go through my forum posts to find out what my details were and check.  I'll let you know what I find out, if I'm not too tired to go check.  :-)

**Comment by sawolf** [comment]
Scott was able to reproduce on a 32-bit machine. The library code for exclusion mapping is dependent on architecture (https://github.com/NagiosEnterprises/nagioscore/blob/master/lib/bitmap.c if you're curious), so I imagine that's probably the issue. I'll step through this again tomorrow.

**Comment by scottwilkerson** [comment]
Digging more and more into this, I could not find the difference between 2 machines that were identical until I realized that on one machine when verifying the configuration it was listing that I had a duplicate service definition for the service I was trying to exclude.

```# /usr/local/nagios/bin/nagios -v /usr/local/nagios/etc/nagios.cfg|grep HTTP
Warning: Duplicate definition found for service 'HTTP' on host 'localhost' (config file '/usr/local/nagios/etc/services/localhost.cfg', starting on line 347)```

When I removed the duplicate definition, the escalation was excluded as expected.

Eric, If you have a different scenerio, I would entertain looking at it, but AFAIK this is working as expected so long as you do not have duplicate service definitions.

**Comment by scottwilkerson** [comment]
Digging more and more into this, I could not find the difference between 2 machines that were identical until I realized that on one machine when verifying the configuration it was listing that I had a duplicate service definition for the service I was trying to exclude.

```
/usr/local/nagios/bin/nagios -v /usr/local/nagios/etc/nagios.cfg|grep HTTP
Warning: Duplicate definition found for service 'HTTP' on host 'localhost' (config file '/usr/local/nagios/etc/services/localhost.cfg', starting on line 347)
```

When I removed the duplicate definition, the escalation was excluded as expected.

Eric, If you have a different scenario, I would entertain looking at it, but AFAIK this is working as expected so long as you do not have duplicate service definitions.

**Comment by scottwilkerson** [comment]
The original forum post is here
https://support.nagios.com/forum/viewtopic.php?f=20&t=55081

**Comment by ericloyd** [comment]
Ah, that's not me.  :-)  No wonder I can't find it in my notes.  Yeah, I have nothing further to add on this one, though I suppose it may be _related_ to one of these I've posted about dependencies:

TASK ID=3948
https://support.nagios.com/forum/viewtopic.php?f=6&t=40187
https://support.nagios.com/forum/viewtopic.php?f=16&t=55267
https://support.nagios.com/forum/viewtopic.php?f=7&t=28200

**Comment by sawolf** [comment]
Okay, so as far as those threads go, I believe the first is #168 and was basically waiting until core 5 (if/when we make configs more generic), and the other two (including task 3948) seem resolved (at least on our end)- 3948 was fixed in Core 4.1.1 (tracked here as #19), the other thread was waiting on more specific configuration info/a minimal example that showed the bad behavior.

Closing the issue now
